### PR TITLE
support `@base/object` module

### PR DESCRIPTION
**Pull request by PPpro**
Re: https://github.com/cocos/cocos-engine/issues/16375

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/asset/asset-manager/asset-manager.ts [hunk 1]
```diff
@@ -29,20 +29,16 @@ import { cclegacy } from '@base/global';
...
-import Cache from './cache';
-import { ICache } from './cache';
+import Cache, { ICache } from './cache';
 import CacheManager from './cache-manager';
-import dependUtil from './depend-util';
-import { DependUtil } from './depend-util';
-import downloader from './downloader';
-import { Downloader } from './downloader';
+import dependUtil, { DependUtil } from './depend-util';
+import downloader, { Downloader } from './downloader';
```
Comment (PPpro) [inline cocos/asset/asset-manager/asset-manager.ts:35]: as it's discussed before https://github.com/cocos/cocos-engine/pull/16247#discussion_r1324651600

we prefer to separate the default import and named import, but currently we have no suitable eslint rule to supporting to do this. 
so I revert this change to adapt to current eslint rules

File: cocos/core/event/callbacks-invoker.ts [hunk 2]
```diff
@@ -26,7 +26,7 @@
 import { TEST } from 'internal:constants';
 import { cclegacy } from '@base/global';
 import { js, memop } from '@base/utils';
-import { isCCObject, isValid } from '../data/object';
+import { isCCObject, isValid } from '@base/object';
```
Comment (PPpro) [inline cocos/core/event/callbacks-invoker.ts:29]: we need this to support `@base/event` module

File: cocos/core/data/index.ts [hunk 3]
```diff
@@ -27,11 +27,7 @@ import * as _decorator from './class-decorator';
 
 cclegacy._decorator = _decorator;
 export { _decorator };
-export { CCClass, isCCClassOrFastDefined } from './class';
-export { CCObject } from './object';
-export { CCInteger, CCFloat, CCBoolean, CCString } from './utils/attribute';
 export { CompactValueTypeArray } from './utils/compact-value-type-array';
-export { editorExtrasTag } from './editor-extras-tag';
```
Comment (PPpro) [inline cocos/core/data/index.ts:34]: a module cannot re-export another module's interfaces

File: cocos/scene-graph/prefab/utils.ts [hunk 4]
```diff
@@ -26,11 +26,10 @@
 import { EDITOR, SUPPORT_JIT } from 'internal:constants';
 import { cclegacy } from '@base/global';
 import { errorID, warn } from '@base/debug';
-import { editorExtrasTag } from '../../core';
+import { editorExtrasTag, ValueType } from '@base/object';
```
Comment (minggo) [inline cocos/scene-graph/prefab/utils.ts:29]: What's `editorExtrasTag `? Why it belongs to object module?

File: cocos/serialization/instantiate.ts [hunk 5]
```diff
@@ -28,7 +28,8 @@ import { getError, warn } from '@base/debug';
 import { cclegacy } from '@base/global';
 import { js } from '@base/utils';
 import { isDomNode } from '@pal/utils';
-import { CCObject, isCCObject, ValueType, jsbUtils, isCCClassOrFastDefined } from '../core';
+import { CCObject, isCCClassOrFastDefined, ValueType, isCCObject } from '@base/object';
```
Comment (minggo) [inline cocos/serialization/instantiate.ts:31]: As `isCCClassOrFastDefined ` is something related with class, not object. It is better to implement in core.

File: cocos/scene-graph/prefab/utils.ts (hunk 4 above)
Comment (PPpro) [reply cocos/scene-graph/prefab/utils.ts:29]: it's a private symbol or string, and this is a js trick to implement `firend` property in js class, for example

```ts
// 'A' module
export const editorExtrasTag = symbol();

export class A {
    // we use compute property to implement friend property in js
    [editorExtrasTag] = 1;
}
```

that is, if any js class can access `editorExtrasTag`, they can access the `[editorExtrasTag]` property in class A

```ts
import { editorExtrasTag } from 'A'
class B {
    a: A = new A();
    accessAFriendProperty () {
        this.a[editorExtrasTag]    // <--
    }
}
```

> Why it belongs to object module?

because `CCObject` implements `EditorExtendableObject`， and `editorExtrasTag` is part of `EditorExtendableObject`

File: cocos/serialization/instantiate.ts (hunk 5 above)
Comment (PPpro) [reply cocos/serialization/instantiate.ts:31]: yes, it should be, but the implementation of `CCObject` rely on `CCClass`, and `isCCClassOrFastDefined` is part of `CCClass`

**Review by minggo — APPROVED** [review]
